_source/_posts/2023-07-27-spring-graphql-react.md

```diff
@@ -0,0 +1,1794 @@
+//Constraint for a node key is a Neo4j Enterprise feature only - run on an instance with enterprise
+//CREATE CONSTRAINT ON (p:Person) ASSERT (p.birthMonth, p.birthYear, p.name) IS NODE KEY
+CREATE CONSTRAINT FOR (p:Property) REQUIRE p.titleNumber IS UNIQUE;
+```
+
+__V002__Company.cypher__
+```cypher
+LOAD CSV WITH HEADERS FROM "file:///PSCAmericans.csv" AS row
+MERGE (c:Company {companyNumber: row.company_number})
+RETURN COUNT(*);
+```
+
+__V003__Person.cypher__
+```cypher
+LOAD CSV WITH HEADERS FROM "file:///PSCAmericans.csv" AS row
+MERGE (p:Person {name: row.`data.name`, birthYear: row.`data.date_of_birth.year`, birthMonth: row.`data.date_of_birth.month`})
+  ON CREATE SET p.nationality = row.`data.nationality`,
+  p.countryOfResidence = row.`data.country_of_residence`
+RETURN COUNT(*);
+```
+
+__V004__PersonCompany.cypher__
+```cypher
+LOAD CSV WITH HEADERS FROM "file:///PSCAmericans.csv" AS row
+MATCH (c:Company {companyNumber: row.company_number})
+MATCH (p:Person {name: row.`data.name`, birthYear: row.`data.date_of_birth.year`, birthMonth: row.`data.date_of_birth.month`})
+MERGE (p)-[r:HAS_CONTROL]->(c)
+SET r.nature = split(replace(replace(replace(row.`data.natures_of_control`, "[",""),"]",""),  '"', ""), ",")
+RETURN COUNT(*);
+```
+
+__V005__CompanyData.cypher__
+```cypher
+LOAD CSV WITH HEADERS FROM "file:///CompanyDataAmericans.csv" AS row
+MATCH (c:Company {companyNumber: row.` CompanyNumber`})
+SET c.name = row.CompanyName,
+c.mortgagesOutstanding = toInteger(row.`Mortgages.NumMortOutstanding`),
+c.incorporationDate = Date(Datetime({epochSeconds: apoc.date.parse(row.IncorporationDate,'s','dd/MM/yyyy')})),
+c.SIC = row.`SICCode.SicText_1`,
+c.countryOfOrigin = row.CountryOfOrigin,
+c.status = row.CompanyStatus,
+c.category = row.CompanyCategory;
+```
+
+__V006__Land.cypher__
+```cypher
+LOAD CSV WITH HEADERS FROM "file:///LandOwnershipAmericans.csv" AS row
+MATCH (c:Company {companyNumber: row.`Company Registration No. (1)`})
+MERGE (p:Property {titleNumber: row.`Title Number`})
+SET p.address = row.`Property Address`,
+p.county  = row.County,
+p.price   = toInteger(row.`Price Paid`),
+p.district = row.District
+MERGE (c)-[r:OWNS]->(p)
+WITH row, c,r,p WHERE row.`Date Proprietor Added` IS NOT NULL
+SET r.date = Date(Datetime({epochSeconds: apoc.date.parse(row.`Date Proprietor Added`,'s','dd-MM-yyyy')}));
+CREATE INDEX FOR (c:Company) ON c.incorporationDate;
+```
+{% endraw %}
+
+Update `application.properties` and add the following properties:
+
+__application.properties__
+```properties
+...
+spring.graphql.graphiql.enabled=true
+spring.graphql.schema.introspection.enabled=true
+org.neo4j.migrations.transaction-mode=PER_STATEMENT
+spring.neo4j.uri=bolt://localhost:7687
+spring.neo4j.authentication.username=neo4j
+
+spring.graphql.cors.allowed-origins=http://localhost:3000
+```
+
+Create a `.env` file in the project root to store the Neo4j credentials:
+
+__.env__
+```shell
+export SPRING_NEO4J_AUTHENTICATION_PASSWORD=verysecret
```

Comment:
I wish Spring Boot supported .env files by default so you don't have to `source` this file first. I recently used [spring-dotenv](https://github.com/paulschwarz/spring-dotenv) for this. 

No need to change to use this, just an FYI.

Reply:
If it was possible to use a dotenv equivalent, you could just have one file in the root directory for Spring Boot and Neo4j in Docker. Then, you could start the Neo4j images with `docker compose -f src/main/docker/neo4j.yml`.

Reply:
In case you weren't aware, Spring Boot 3.1 has Docker Compose support built-in. This means you can add a dependency and a compose file in the root directory, and Spring Boot will auto-start it. https://spring.io/blog/2023/06/21/docker-compose-support-in-spring-boot-3-1

I used it in this example I created last week: https://github.com/oktadev/auth0-java-microservices-examples/tree/main/spring-boot-gateway-webflux/car-service